libhail/src/hail/query/backend/stype.hpp

```diff
@@ -77,6 +77,18 @@ class SFloat64 : public SType {
   SFloat64(const Type *type);
 };
 
+class SArray : public SType {
+  public:
+    static const Tag self_tag = SType::Tag::ARRAY;
```

Comment:
This won't work as it stands.  Only concrete members of the hierarchy can have tags, so we shouldn't have this, and there should be a `Tag::CANONICAL_ARRAY`.

There is a question if we should have SArray at all.  If you want to be able to do `isa<SArray>(stype)` or `cast<SArray>(stype)`, then we enhance the casting stuff to support non-concrete intermediate classes (I would probably avoid concrete intermediate classes for now).